### PR TITLE
Feature/api for models with permissions

### PR DESCRIPTION
**Pull request by ArtsiomSt**
If you see that a lot files were changed, that is because my system changed permissions for files in project for IDE, so i had to change permissions by chmod and git catched this change

### CODE REVIEW

File: CarShowRoom/cars/api/permissions.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from rest_framework.permissions import SAFE_METHODS, BasePermission
+
+
+class IsDealerOrReadOnly(BasePermission):
+    def has_permission(self, request, view):
+        if not request.user.is_authenticated:
+            return False
+        return bool(
+            request.user.user_type == "DEALER"
```
Comment (rupeq) [inline CarShowRoom/cars/api/permissions.py]: use enum: `UserType.DEALER.name`

File: CarShowRoom/cars/api/v1/views.py [hunk 2]
```diff
@@ -59,3 +67,37 @@ def update(self, request, *args, **kwargs):
...
+    def delete(self, request, car_id):
+        dealer_with_car_id = DealerCar.objects.filter(
+            dealer=request.user.id, car=car_id
+        )
+        if dealer_with_car_id:
+            dealer_with_car_id.delete()
+            return return_message("Car has been successfully deleted from your list")
+        return return_message(
+            "there is no such car in your list", status.HTTP_400_BAD_REQUEST
+        )
```
Comment (rupeq) [inline CarShowRoom/cars/api/v1/views.py]: I don't think that should be a separate view (but still can be). We can add extra endpoint for either car view or dealer view using @actions decorator.

Comment (rupeq) [inline CarShowRoom/cars/api/v1/views.py]: we can also write our custom DynamicSerializerMixin and DynamicPermissionMixin classes.
then we can inherit our views and define dicts in view class body:


```
 serializer_mapping = {
        ("list"): CarSerializer,
        ("update", "create"): CarActionSerializer,
        ("get_dealer_cars): DealerCarSerializer
 }

 permission_mapping = {
        ("update", "create", "get_dealer_cars"): IsDealerOrReadOnly | IsSuperUser)
 }
```


dicts define action tuple as a key and permission classes as a value. then our mixins check
lmk if you have any questions

File: CarShowRoom/core/serializers.py [hunk 3]
```diff
@@ -22,6 +24,8 @@ class RegisterSerializer(serializers.ModelSerializer):
     )
 
     def create(self, validated_data):
+        if hasattr(self, "Meta"):
+            validated_data["user_type"] = self.Meta.model.__name__.upper()
```
Comment (rupeq) [inline CarShowRoom/core/serializers.py:28]: what for?

File: CarShowRoom/sellers/api/permissions.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+from rest_framework.permissions import SAFE_METHODS, BasePermission
+
+
+class IsShowRoomOwnerOrReadOnly(BasePermission):
+    def has_permission(self, request, view):
+        request_kwargs = request.parser_context["kwargs"]
+        if not request.user.is_authenticated:
+            return False
+        return bool(
+            request.user.user_type == "CARSHOWROOM"
```
Comment (rupeq) [inline CarShowRoom/sellers/api/permissions.py]: same

File: CarShowRoom/customers/tests/fixtures.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+@pytest.fixture()
+def customer_without_email(password):
+    return Customer.objects.create(
+        first_name="customer_test",
+        last_name="customer_test",
+        username="customer_no_email",
+        email="customer_test_no_email@testpytest.com",
+        password=password,
+        is_email_verified=False,
+    )
```
Comment (rupeq) [inline CarShowRoom/customers/tests/fixtures.py]: you can look at test factories technique (e.g. https://schegel.net/posts/simplied-django-tests-with-pytest-and-pytest-factoryboy/)

**Review by rupeq — CHANGES_REQUESTED** [review]

File: CarShowRoom/core/serializers.py (hunk 3 above)
Comment (ArtsiomSt) [reply CarShowRoom/core/serializers.py:28]: It is used to automatically fill user_type field in models, choices in user types are similar with models names so I used this way to fulfill this field

File: CarShowRoom/cars/api/v1/views.py (hunk 2 above)
Comment (ArtsiomSt) [reply CarShowRoom/cars/api/v1/views.py]: Yeah agree, i checked this way, i will integrate this view into carviewset)

**Comment by ArtsiomSt** [comment]
I have created DynamicSerializerMixin, but i doubt the way to get the name of the method from which we want to get the serializer(i have used call-stack), maybe you can suggest a better way to find parent-method name? Currently i dont have cases where i need  DynamicPermissionMixin, so i will try to implement this mixin tomorrow

File: CarShowRoom/cars/api/v1/urls.py [hunk 6]
```diff
@@ -7,3 +8,16 @@
 router.register(r"car", CarViewSet)
 
 urlpatterns = router.urls
+
+urlpatterns += [
+    path(
+        "car/<int:pk>/add/",
```
Comment (rupeq) [inline CarShowRoom/cars/api/v1/urls.py]: i don't think you need to have /add/ here, maybe car/id/dealer or something ?

File: CarShowRoom/core/mixins.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        assert self.serializer_mapping is not None
+        for methods, serializer_class in self.serializer_mapping.items():
+            called_from = stack()[1].function
+            if called_from != "get_serializer":
+                raise ValueError("get_serializer_class should be called from get_serializer")
+            for depth in range(len(stack())):
+                function_of_that_level = stack()[depth].function
+                if function_of_that_level in methods:
+                    return serializer_class
+        raise ValueError("Bad serializer mapping")
```
Comment (rupeq) [inline CarShowRoom/core/mixins.py]: I would define serializer mappin as empty dict  


 ```
 def get_serializer_class(self):
        for action, serializer in self.serializer_mapping.items():
            if self.action == action or self.action in action:
                return serializer
        return self.serializer_class
```

**Review by rupeq — CHANGES_REQUESTED** [review]

File: CarShowRoom/core/mixins.py (hunk 7 above)
Comment (rupeq) [reply CarShowRoom/core/mixins.py]: where `self.serializer_class` is a default serializer

Comment (ArtsiomSt) [reply CarShowRoom/core/mixins.py]: Didnt know there is an action attribute, will fix

File: CarShowRoom/cars/api/v1/urls.py [hunk 8]
```diff
@@ -7,3 +8,16 @@
 router.register(r"car", CarViewSet)
 
 urlpatterns = router.urls
+
+urlpatterns += [
+    path(
+        "car/<int:pk>/managedealercars/",
```
Comment (rupeq) [inline CarShowRoom/cars/api/v1/urls.py:14]: Nit: usually endpoint name is a corresponding model name. If endpoint is mixed (e.g a single model name cannot be specified), then you can combine into one using hyphens 

https://restfulapi.net/resource-naming/

**Review by rupeq — APPROVED** [review]
Lgtm 1nit

File: CarShowRoom/cars/api/v1/urls.py (hunk 8 above)
Comment (ArtsiomSt) [reply CarShowRoom/cars/api/v1/urls.py:14]: Okey, will check it

**Pull request merged**